### PR TITLE
StormCast recipe support for downscaling/nowcasting/unconditional models

### PR DESCRIPTION
**Pull request by jleinonen**
# PhysicsNeMo Pull Request

Based partially on feedback by @mariusaurus, this PR ensures that different types of regional forecasting models can be trained with the StormCast recipe and makes their implementation smoother (e.g. by removing the need to use dummy tensors in `__getitem__`).

Changes:

1. Add a test in `test_training.py` to train with different model configurations (downscaling, nowcasting, StormCast-like hybrid models and unconditional diffusion models).
2. Fix various small issues identified by the testing that prevented some configurations from training.
3. Eliminate the need to pass dummy tensors from `__getitem__`: `"background"` may be omitted by nowcasting models that do not use low-resolution conditioning, while `state` may be a `torch.Tensor` rather than a list of two tensors for models that do not do a state update (downscaling and unconditional models).
4. Update `README.md` with more information about configuring the recipe to train different types of models.
5. Update `requirements.txt` and `README.md` to note that torch>=2.10 is needed for domain parallelism.
6. Various other minor improvements to `README.md`.
7. Fix `ConcatConditionWrapper` where an `or` in an error check should have been `and`.
8. Remove StableAdamW support as it doesn't work properly with FSDP even when domain parallelism is not used.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: examples/weather/stormcast/utils/sampler.py [hunk 1]
```diff
@@ -272,7 +272,8 @@ def deterministic_sampler(
                 f"mean_hr and img_lr must have the same height and width, "
                 f"but found {mean_hr.shape[-2:]} vs {img_lr.shape[-2:]}."
             )
-        x_lr = torch.cat((mean_hr.expand(x_lr.shape[0], -1, -1, -1), x_lr), dim=1)
+        mean_hr = mean_hr.expand(x_lr.shape[0], -1, -1, -1)
+        x_lr = torch.cat((mean_hr, x_lr), dim=1) if x_lr is not None else mean_hr
```
[the remark this answers is not in the document]
Comment (jleinonen) [reply examples/weather/stormcast/utils/sampler.py]: Fixed in https://github.com/NVIDIA/physicsnemo/pull/1474/commits/e032f9ab2571e932aaa92d215b30ef7e213d1892

File: examples/weather/stormcast/utils/loss.py [hunk 2]
```diff
@@ -165,7 +165,7 @@ def __call__(
                 **optional_args,
             )
         else:
-            D_yn = net(y + n, sigma, labels, **optional_args)
+            D_yn = net(y + n, sigma.flatten(), labels, **optional_args)
```
Comment (pzharrington) [inline examples/weather/stormcast/utils/loss.py:168]: What was the reasoning for this `flatten()`?

File: examples/weather/stormcast/README.md [hunk 3]
```diff
@@ -1,19 +1,19 @@
 <!-- markdownlint-disable -->
-# StormCast: Kilometer-Scale Convection Allowing Model Emulation using Generative Diffusion Modeling
+# StormCast: Regional High-Resolution Weather Models using Generative Diffusion
```
Comment (pzharrington) [inline examples/weather/stormcast/README.md]: I suggest we drop the `StormCast` from this title entirely. Then soon we can migrate it to a `regional_generative_models` example officially 🚀

**Review by pzharrington — APPROVED** [review]
Awesome changes!

File: examples/weather/stormcast/utils/loss.py (hunk 2 above)
Comment (jleinonen) [reply examples/weather/stormcast/utils/loss.py:168]: `EDMPreconditioner` (the new class) requires that `sigma` has a 1D `(B,)` shape. Without the `.flatten()`, this caused an error in the specific case where we have an unconditional model (which selects the `else` branch here) and a `DiT` (which uses the new `EDMPreconditioner`).  The added `.flatten()` fixes that and still seems to work ok with an unconditional model in combination with a `SongUNet`.

File: examples/weather/stormcast/README.md (hunk 3 above)
Comment (jleinonen) [reply examples/weather/stormcast/README.md]: Done in https://github.com/NVIDIA/physicsnemo/pull/1474/commits/677aa419e5fe4f6f5afa2d7c542b2687cc115c96.

File: examples/weather/stormcast/datasets/mock.py [hunk 4]
```diff
@@ -68,10 +72,20 @@ def __getitem__(self, idx: int) -> dict[str, Any]:
...
+            }
+        elif self._model_type == "nowcasting":
+            item = {"state": [state_input, state_target]}
+        elif self._model_type == "downscaling":
+            item = {
+                "background": background,
+                "state": state_target,
+            }
+        elif self._model_type == "unconditional":
+            item = {"state": state_target}
```
[the remark this answers is not in the document]
Comment (jleinonen) [reply examples/weather/stormcast/datasets/mock.py:88]: Fixed in https://github.com/NVIDIA/physicsnemo/pull/1474/commits/d0a08b14a105c52564174b7646bbb262e7dc468e.
